src/DominantJuice.sol

```diff
@@ -0,0 +1,399 @@
+     * to receive a `didPay()` call. `JBPayoutRedemptionPaymentTerminal3_1_1.pay()` calls
+     * `JBSingleTokenPaymentTerminalStore3_1_1.recordPaymentFrom()`, which calls this function.
+     * @param _data JB project payment data. See https://docs.juicebox.money/dev/api/data-structures/jbpayparamsdata/.
+     * @return weight The weight that project tokens should get minted relative to. This is useful for optionally
+     * customizing how many tokens are issued per payment.
+     * @return memo A memo to be forwarded to the event. Useful for describing any new actions that are being taken.
+     * @return delegateAllocations Amount to be sent to delegates instead of adding to local balance. Useful for
+     * auto-routing funds from a treasury as payment come in.
+     */
+    function payParams(JBPayParamsData calldata _data)
+        external
+        view
+        virtual
+        override
+        returns (uint256 weight, string memory memo, JBPayDelegateAllocation3_1_1[] memory delegateAllocations)
+    {
+        // The if statements are the only changes from the payParams() template. This is also not a payable
+        // function, so calls with msg.value will revert.
+        if (getBalance() == 0) revert RefundBonusNotDeposited();
+        if (hasCycleExpired()) revert CycleHasExpired();
+        if (_data.amount.value < campaign.minimumPledgeAmount) {
+            revert AmountIsBelowMinimumPledge(campaign.minimumPledgeAmount);
+        }
+
+        // Forward the default weight received from the protocol.
+        weight = _data.weight;
+        // Forward the default memo received from the payer.
+        memo = _data.memo;
+        // Add `this` contract as a Pay Delegate so that it receives a `didPay` call. Don't send funds to the delegate (keep in treasury).
+        delegateAllocations = new JBPayDelegateAllocation3_1_1[](1);
+        delegateAllocations[0] = JBPayDelegateAllocation3_1_1(this, 0, "");
+    }
+
+    /**
+     * @notice Received hook from the payment terminal after a payment.
+     * @dev Reverts if the calling contract is not one of the project's terminals. This is the final call in the
+     * chain described in `payParams()`.
+     * @param _data Standard Juicebox project payment data.
+     * See https://docs.juicebox.money/dev/api/data-structures/jbdidpaydata3_1_1/.
+     */
+    function didPay(JBDidPayData3_1_1 calldata _data) external payable virtual override terminalCheck {
+        if (_data.projectId != campaign.projectId) revert IncorrectProjectID();
+
+        // Get payer address and amount paid from JB DidPay data struct
+        address payer = _data.payer;
+        JBTokenAmount memory amount = _data.amount;
+        uint256 paymentAmount = amount.value;
+
+        // Get time of pledge from campaign start.
+        (JBFundingCycle memory cycleData,) = controller.currentFundingCycleOf(campaign.projectId);
+        uint256 hourOfPledge = (block.timestamp - cycleData.start) / 3600;
```

Comment:
🧪 make sure there's a test for calling this after the campaign is done

Reply:
There was one at the bottom of that function’s section. I’ll update it as well for the new lockPeriod storage variable.

Reply:
ah, ok. I read this before the tests and forgot to check it off